### PR TITLE
Ease testing of simple rule standard cases

### PR DESCRIPTION
**Pull request by kylekatarnls**
Type: refactoring
Breaking change: no

This abstraction provides some intermediate methods to better customize node selection for rule testing and provide a way to test all sample files for basic applies/doesNotApply test cases.

The result is demonstrated in this PR on UndefinedVariable rule.

### CODE REVIEW

**Comment by kylekatarnls** [comment]
For the double-check: unit tests count of this PR is still 554 as in the last Travis build of the master branch.

**Comment by ravage84** [comment]
Thank you for the proposal.

This is more or less what I expected.

For me, this is a too extreme abstraction.
The tests becomes __too__ simple, even primitive.

And on top of that, instead of improving the test readability and expressiveness, it reduces them by introduces even more magic than what we had before with ``$rule->apply($this->getMethod())``.

```` php
/**
 * Test case for the undefined variable rule.
 *
 * @covers \PHPMD\Rule\CleanCode\UndefinedVariable
 * @covers \PHPMD\Rule\AbstractLocalVariable
 */
class UndefinedVariableTest extends SimpleRuleTestCase
{
    public function getRule()
    {
        return new UndefinedVariable();
    }
}
````

But yes, with that we would write less redundant test code.

**Comment by kylekatarnls** [comment]
> The tests becomes too simple, even primitive

The tests are the same as before. But actually result for new contributors is:

Before:
1. Add tests sample files
2. Guess the magic naming within the file (follow other files example)
3. Guess the magic method naming: add the method with the same name in the test case class

After:
1 + 2, but you get rid of 3 and so have less magic linking to retro-engineer.

**Comment by kylekatarnls** [comment]
An other point to consider, is you find it "too" simple while I think it already is simple test (for this file) but the big methods count give a false impression of doing different tests while we actually do only 1 kind of test. Now, that we have this simple test we actually can focus on edge cases:

```php
/**
 * Test case for the undefined variable rule.
 *
 * @covers \PHPMD\Rule\CleanCode\UndefinedVariable
 * @covers \PHPMD\Rule\AbstractLocalVariable
 */
class UndefinedVariableTest extends SimpleRuleTestCase
{
    public function getRule()
    {
        return new UndefinedVariable();
    }

    public function testEdgeCaseFoo()
    {
        // Really specific test
    }

    public function testEdgeCaseBar()
    {
        // Really specific test
    }
}
```

The new system does not prevent from adding tests. And those tests can use resources files too, it just need to have a name that does not match the "testRuleApplyTo" pattern.

So now let's imagine all test case classes could be transformed like so, you would have a clear view of standard/specific tests. And adding tests would not be less obvious than before, on the contrary.

**Comment by ravage84** [comment]
I was gonig to create an alternative PR but how about this:

```` php
<?php
class UndefinedVariableTest extends AbstractTestCase
{
    public function testRuleAppliesTo()
    {
        // All normal cases where the rule applies, using the data provider below
    }

    public function dataProviderCasesThatApply()
    {
        return array(
            'testRuleAppliesToUndefinedVariable',
            'testRuleAppliesToUndefinedVariableWithDefinedVariable',
            'testRuleAppliesToUndefinedVariableOnArray',
            'testRuleAppliesToUndefinedVariableOnArrayWithKeys',
            'testRuleAppliesToUndefinedVariable',
        );
    }

    public function testRuleDoesNotApplyTo()
    {
        // All normal cases where the rule does not apply, using the data provider below
    }

    public function dataProviderCasesThatDoNotApply()
    {
        return array(
            'testRuleDoesNotApplyToSuperGlobals',
            'testRuleDoesNotApplyToUsedProperties',
        );
    }

    public function testRuleAppliesToWithPropertyXyz()
    {
        // All cases where the rule applies in connection with the property Xyz, using the data provider below
    }

    public function dataProviderPropertyXCasesThatApply()
    {
        return array(
            'testRuleApplyToWithPropertyXFoo',
            'testRuleApplyToWithPropertyXBar',
        );
    }

    public function testRuleDoesNotApplyToWithPropertyXyz()
    {
        // All cases where the rule does not in connection with the property Xyz, using the data provider below
    }

    public function dataProviderPropertyXCasesThatDoNotApply()
    {
        return array(
            'testRuleDoesNotApplyToWithPropertyXFoo',
            'testRuleDoesNotApplyToWithPropertyXBar',
        );
    }

    public function testEdgeCaseFoo()
    {
        // Really specific test case with no data provider
    }

    public function testEdgeCaseBar()
    {
        // Really specific test case with no data provider
    }
}
````
This way we reduce the amount of test case methods, we group together what belongs together, keep or even increas the readability

But yes, this is still more code than your solution, but that is not a problem.

**Comment by kylekatarnls** [comment]
Actually this PR allows to do what you propose. For example `testRuleAppliesTo` is `testRuleAppliesToSuccessFiles` in `SimpleRuleTestCase` so you can override it if your test content is not the standard thing (class method test).

Then `dataProviderCasesThatDoNotApply` is `getSuccessFiles` in `SimpleRuleTestCase` so you can override it if your test files list is not the standard pattern.

Advantage of not overriding `SimpleRuleTestCase` is still to give one less step for contributor of finding the dataProvider to fill.

About "dataProvider" naming prefix, I really dislike to name methods about where/how/why they are called. Every method should be named only according to what they do no matter the calls.

**Comment by ravage84** [comment]
> About "dataProvider" naming prefix, I really dislike to name methods about where/how/why they are called. Every method should be named only according to what they do no matter the calls.

I am OK with not using that prefix/suffix.
Just used it for more clarity in the example.

**Comment by kylekatarnls** [comment]
`getRule()` And `SimpleRuleTestCase` are not good naming regarding the possibilities, I will change that first so this PR could become more clear.

**Comment by kylekatarnls** [comment]
In this new version, dataProvider is no longer hidden behind the inherited class:
https://github.com/phpmd/phpmd/blob/5295875e42193d368fceaa605f99d31917367324/src/test/php/PHPMD/Rule/CleanCode/UndefinedValiableTest.php

`SimpleRuleTestCase` disappeared, its methods are directly provided by `AbstractTest` as conveniant getters of test files with the PHPMD naminig convention.

Then the actual test methods are inside the sub-class and you can easily go to the `@dataProvider` definition with your IDE or override it on need.

File: src/test/php/PHPMD/Rule/CleanCode/UndefinedValiableTest.php [hunk 1]
```diff
@@ -28,74 +28,26 @@
...
-     * @return void
+     * @dataProvider getSuccessCases
      */
-    public function testRuleAppliesToUndefinedVariableWithDefinedVariable()
+    public function testRuleAppliesToSuccessFiles($file)
     {
-        $rule = new UndefinedVariable();
-        $rule->setReport($this->getReportMock(1));
-        $rule->apply($this->getMethod());
+        $this->expectRuleInvokesForFile($this->getRule(), 1, $file);
```
Comment (kylekatarnls) [inline src/test/php/PHPMD/Rule/CleanCode/UndefinedValiableTest.php:43]: Here would goes [ONE_VIOLATION constant](https://github.com/phpmd/phpmd/pull/757/files#diff-a53e0ba74d6ad5dc469e51319958ea8dR48) of @ravage84 PR #757

File: src/test/php/PHPMD/Rule/CleanCode/UndefinedValiableTest.php [hunk 2]
```diff
@@ -28,74 +28,26 @@
...
-     * testRuleDoesNotApplyToUsedProperties
-     *
-     * @return void
-     */
-    public function testRuleDoesNotApplyToUsedProperties()
-    {
-        $rule = new UndefinedVariable();
-        $rule->setReport($this->getReportMock(0));
-        $rule->apply($this->getMethod());
+        $this->expectRuleInvokesForFile($this->getRule(), 0, $file);
```
Comment (kylekatarnls) [inline src/test/php/PHPMD/Rule/CleanCode/UndefinedValiableTest.php:51]: Here would goes [NO_VIOLATION constant](https://github.com/phpmd/phpmd/pull/757/files#diff-a53e0ba74d6ad5dc469e51319958ea8dR45) of @ravage84 PR #757

File: src/test/php/PHPMD/AbstractStaticTest.php [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+ * Abstract base class for PHPMD test cases.
+ */
+abstract class AbstractStaticTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * Directory with test files.
+     *
+     * @var string $_filesDirectory
+     */
+    private static $filesDirectory = null;
```
Comment (ravage84) [inline src/test/php/PHPMD/AbstractStaticTest.php:30]: Can we rebase this PR on https://github.com/phpmd/phpmd/pull/758, please?

File: src/test/php/PHPMD/AbstractTest.php [hunk 4]
```diff
@@ -155,124 +152,93 @@ protected function getFunction()
...
      *
-     * @param string $localPath The local/relative file location
-     * @return string
-     * @since 1.1.0
+     * @param Rule   $rule            Rule to test.
+     * @param int    $expectedInvokes Count of expected invocations.
+     * @param string $file            Test file containing a method with the same name to be tested.
      */
-    protected static function createResourceUriForTest($localPath)
+    protected function expectRuleInvokesForFile(Rule $rule, $expectedInvokes, $file)
```
Comment (ravage84) [inline src/test/php/PHPMD/AbstractTest.php:181]: How about ``expectRuleInvokesNTimesForFile``, instead?
Or may be even better: ``expectRuleHasViolationsForFile(Rule $rule, $expectedViolationCount, $file)``

File: src/test/php/PHPMD/AbstractStaticTest.php [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+ * @license https://opensource.org/licenses/bsd-license.php BSD License
+ * @link http://phpmd.org/
+ */
+
+namespace PHPMD;
+
+/**
+ * Abstract base class for PHPMD test cases.
+ */
+abstract class AbstractStaticTest extends \PHPUnit_Framework_TestCase
```
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractStaticTest.php:23]: If by composition, you mean traits, nope we can't (PHP 5.4)

File: src/test/php/PHPMD/AbstractStaticTest.php (hunk 3 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractStaticTest.php:30]: The current merge, seem broken, I will force-push my branch.

Comment (ravage84) [reply src/test/php/PHPMD/AbstractStaticTest.php:30]: No, I thought about simly calling it through static access, like ``AbstractStaticTest::doSomething()``. We should rename the class, though.

File: src/test/php/PHPMD/AbstractStaticTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (ravage84) [reply src/test/php/PHPMD/AbstractStaticTest.php:23]: No, I thought about simly calling it through static access, like ``AbstractStaticTest::doSomething()``. We should rename the class, though.

File: src/test/php/PHPMD/AbstractTest.php [hunk 6]
```diff
@@ -39,30 +39,18 @@
...
-        return $this->getFilesForCalledClass('testRuleAppliesTo*');
-    }
-
-    /**
-     * @return string[]
-     */
-    public function getFailureFiles()
-    {
-        return $this->getFilesForCalledClass('testRuleDoesNotApplyTo*');
-    }
```
Comment (kylekatarnls) [inline src/test/php/PHPMD/AbstractTest.php]: Nope, this is for override purpose, it Allow to customize as you propose earlier the files list:

A particular test case class can do:
```php
public function getFailureFiles()
{
  return array(
    'file1',
    'file2',
  );
}
```
So it can add files that are not in the base naming convention or remove some to run a particular test for it.

File: src/test/php/PHPMD/AbstractTest.php [hunk 7]
```diff
@@ -39,14 +39,24 @@
  */
 abstract class AbstractTest extends AbstractStaticTest
 {
-    public function getSuccessCases()
+    /**
+     * Get a list of test files that are named like "testRuleAppliesTo*"
+     *
+     * @return mixed[]
```
Comment (kylekatarnls) [inline src/test/php/PHPMD/AbstractTest.php:45]: @ravage84 It's rather string[][]

Comment (ravage84) [reply src/test/php/PHPMD/AbstractTest.php:45]: In practise, yes. But by signature, it could be anything...

**Comment by kylekatarnls** [comment]
I fixed the conflict, re-applied your changes and use your new constants.

File: src/test/php/PHPMD/AbstractTest.php (hunk 7 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractTest.php]: It's always an array of array and anything else but string will fail, so as guidance for sub-class override, we should force this typing.

Comment (ravage84) [reply src/test/php/PHPMD/AbstractTest.php]: What do you mean by "force"?

File: src/test/php/PHPMD/RuleSetTest.php [hunk 8]
```diff
@@ -85,16 +85,16 @@ public function testApplyInvokesRuleWhenStrictModeIsSet()
...
      * objects.
      *
-     * @return \PHPMD\AbstractRule
+     * @return RuleSet
      */
     private function createRuleSetFixture()
     {
         $ruleSet = new RuleSet();
-        for ($i = 0; $i < func_num_args(); ++$i) {
-            $rule = new RuleStub(func_get_arg($i));
```
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/test/php/PHPMD/RuleSetTest.php:94]: Variable is available for debugging as `addRule(Rule $rule)`, step in, check `$rule`, step out.

I'm OK for intermediate variable if they have a meaningful name, but when having a redundant name like $rule (instance of `Rule` interface) passed to `addRule` of the `$ruleSet`. 1 more "rule" does not add much understanding/readability value.

"Add a rule rule to my set" is not more readable than "Add a rule to my set".

Or in code:
```php
$undefinedVariableRule = new RuleStub();
$ruleSet->addRule($undefinedVariableRule)
```

Yeah, here is a great intermediate variable that gives me information.

But if the variable name bring no information, just skip the noisy line and get straight to the point.

File: src/test/php/PHPMD/AbstractTest.php (hunk 6 above)
Comment (ravage84) [reply src/test/php/PHPMD/AbstractTest.php]: Oh, I'm sorry. The intent wasn't clear. Can you please add an explanation to the doc block? So people understand what this is meant for.

File: src/test/php/PHPMD/AbstractStaticTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractStaticTest.php:23]: Most of them are assert* which are rather usually called with `$this->` or `static::` and there is `setUpBeforeClass` that is auto-called by name by PHPUnit.

They are test case tooling that does not require the `$this` context.

File: src/test/php/PHPMD/AbstractStaticTest.php (hunk 3 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractStaticTest.php:30]: #758 is not merged, it would not make sense to rebase/merge it if it's actually not merged first. So let's wait for one to be merged, then we'll fix the conflict in the other one.

File: src/test/php/PHPMD/AbstractTest.php (hunk 6 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractTest.php]: Added.

File: src/test/php/PHPMD/AbstractStaticTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (ravage84) [reply src/test/php/PHPMD/AbstractStaticTest.php:23]: Well, I guess AbstractTest to some degree and AbstractStaticTest do more than they should, which leads to so many methods on them. May be we should refactor them. But let's look at them in another PR.

File: src/test/php/PHPMD/AbstractTest.php [hunk 9]
```diff
@@ -49,8 +49,9 @@ abstract class AbstractTest extends AbstractStaticTest
     const ONE_VIOLATION = 1;
 
     /**
-     * List files that should trigger a rule violation.
-     * By default, files named like "testRuleAppliesTo*", but it can be overridden in sub-class.
+     * Get a list of files that should trigger a rule violation
+     *
+     * By default, files named like "testRuleAppliesTo*", but it can be overridden in a sub-class.
```
Comment (kylekatarnls) [inline src/test/php/PHPMD/AbstractTest.php:54]: "a sub-class" is not correct, if all our tests we always extend AbstractTest by only 1 level of inheritance. So either you consider this read from the AbstractTest point of view and it can be overridden in the sub-classes (meaning each of the sub-class can override it or not, not exclusively) a (as 1) would make think only 1 can.

Or you consider this from the contributor point of view that extend this class: "You can override it in your class extending AbstractTest, e.g. here in your MyNewRuleTest class."

So "in sub-class" (en sous-classe in French).

See:
https://translate.google.com/#view=home&op=translate&sl=fr&tl=en&text=en%20sous-classe
https://translate.google.com/#view=home&op=translate&sl=en&tl=de&text=in%20subclass

It sounds like the "der" is mandatory in German, while it's actually not in English or French, when speaking of a kind of place, but not one particular place.

I may ask tomorrow to an English colleague.

File: src/test/php/PHPMD/AbstractTest.php [hunk 10]
```diff
@@ -49,9 +49,9 @@ abstract class AbstractTest extends AbstractStaticTest
     const ONE_VIOLATION = 1;
 
     /**
-     * Get a list of files that should trigger a rule violation
+     * Get a list of files that should trigger a rule violation.
```
Comment (kylekatarnls) [inline src/test/php/PHPMD/AbstractTest.php:52]: Dots should be preserved: because it's a sentence. And if this PHPDoc is extracted somewhere, extractors may assume new lines are just there because of the line length maximum and should not be kept for a proper display as text/HTML or in the IDE tooltip. In such case, the dot will help to understand the cut.

Comment (ravage84) [reply src/test/php/PHPMD/AbstractTest.php:52]: > Dots should be preserved: because it's a sentence. 

As explained in some other isssue already: This is by definition of the doc blocks a short description. It can be understood as title and should be brief.

> The summary is a short but effective overview of an element; you can compare it to a slogan or _headline_. 

https://docs.phpdoc.org/latest/guides/docblocks.html#summary

The first variant has become the de-facto standard over the years.

> extractors may assume new lines are just there because of the line length maximum and should not be kept for a proper display as text/HTML or in the IDE tooltip. In such case, the dot will help to understand the cut.

The standard defines a white line (which also improves readability, btw) and in that case a period is superfluous.

If you _really_ insist on a period, then add one :roll_eyes:

File: src/test/php/PHPMD/AbstractTest.php (hunk 9 above)
Comment (ravage84) [reply src/test/php/PHPMD/AbstractTest.php]: > "a sub-class" is not correct, 

@kylekatarnls it is. But it's more a question of style.

In German you either use "ein" (a, as in "any") or der (the) to distinguish between a clearly particular thing (the/der) or any of its kind (a/eine).

But this is not bound to German. It's about writing style. To my knowledge, yes, in English you __may__ omit it but me coming from a precise language like German, I tend to add it for clarity.

Since in this case we do not know the particular sub-class, we cannot speak of it. Thus we generalize and use a/eine sub-class.

"your sub-class" might be more appropriate but I tend to shy away to "directly speak" with the developer.

> "A" (as 1) would make think only 1 can.

Then you would write "one", not "a".

@MarkVaughn can you help us on that?

Comment (ravage84) [reply src/test/php/PHPMD/AbstractTest.php]: How about ``in sub-classes"? Problem solved :smirk_cat:

File: src/test/php/PHPMD/AbstractTest.php (hunk 10 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractTest.php:52]: PHPDoc is documentation, it's text, it should use English punctuation. Headlines have periods in English. "Return something." is an imperative sentence.

This does not prevent to use the formatting as described in phpdoc.org. AS the formatting does not prevent to respect the punctuation.

File: src/test/php/PHPMD/AbstractTest.php (hunk 9 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractTest.php]: I don't say "a sub-class" can't be used, I mean "in sub-class" does not mean the same than "in a sub-class". I'm not a native speaker but yes "in sub-classes" is probably more correct and does not change the sense.

File: src/test/php/PHPMD/AbstractStaticTest.php [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+                    PATH_SEPARATOR,
+                    self::$filesDirectory,
+                    PATH_SEPARATOR,
+                    realpath(__DIR__ . '/../')
+                )
+            );
+        }
+
+        // Prevent timezone warnings if no default TZ is set (PHP > 5.1.0)
+        date_default_timezone_set('UTC');
```
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractStaticTest.php:205]: This code has not been added, just moved. In unit tests, you should use predetermined settings rather than working with random OS settings.

Then as PHPMD is not responsible of date stuff, it's fair to use UTC for tests.

Side note: always use UTC as default timezone in your applications and drive this at application-level, don't let the environment/OS drive the default timezone.

File: src/test/php/PHPMD/AbstractStaticTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractStaticTest.php]: Yeah and AbstractTest has a "red" overall complexity. In fact if I would in PHP >= 5.4, this would have been 5/6 small traits named regarding the methods purposes. But as in PHP 5.3 level we should also avoid a long inheritance chaining. So for this smaller split.

File: src/test/php/PHPMD/AbstractStaticTest.php (hunk 3 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractStaticTest.php:32]: Merged.

**Comment by kylekatarnls** [comment]
Master merged, please use "squash" if you merge it, as it's too many noisy commits for what it does.

File: src/test/php/PHPMD/AbstractStaticTest.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (tvbeek) [reply src/test/php/PHPMD/AbstractStaticTest.php:205]: Thanks, I did a search but suspect that the diff where it was removed was collapsed while searching.

**Comment by kylekatarnls** [comment]
New conflict resolved.
